### ISSUE TITLE
Influxdb v2.7.1 OSS  "Strict-Transport-Security" header.

### ISSUE THREAD

**Issue by 3BK** [influxdata/influxdb#24243]
# Question
Is it just me, or is Influxdb v2.7.1 OSS NOT returning "Strict-Transport-Security" header. when "hardening-enabled : true"?

List the minimal actions needed to reproduce the behavior.

1.  Install inflxudb v2 (latest) OSS
2.  Enable TLS with hardening-enabled
3. Using curl, python,  wget, wireshark, fiddler, openscap, whatever, connect to the influx server and trace the connection sequence

__Expected behavior:__
Expected to see Strict-Transport-Security “max-age=31536000; includeSubDomains” 

__Actual behavior:__
No sign of Strict-Transport-Security “max-age=31536000; includeSubDomains”

__Environment info:__
* System info: Upon request
* InfluxDB version: InfluxDB v2.7.1 (git: 407fa622e9) build_date: 2023-04-28T13:24:27Z

__Config:__
* upon request.

__Logs:__
* Upon request.

__Performance:__
* Upon request.

**Comment by 3BK** [comment]
By way of background: 
(Strict-Transport-Security)[https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Strict-Transport-Security]

**Comment by jdstrand** [comment]
Thank you for the report.

> * Question:
>   * how does Influxdb v2 ensure that the issue above doesn't violate their claim that the instance is secure?

HSTS is meant to help with situations where two ports are being used, one that serves HTTP on 80/tcp and one that serves HTTPS on 443/tcp such that when HSTS is configured, IF the browser at some point went to port 443/tcp, the header is set and the browser records this so that future attempts to port 80/tcp will be converted to requests to port 443/tcp internally by the browser without a redirect. This helps prevent downgrade attacks and situations where the site is serving content over HTTP and, eg, some old page has a direct HTTP link to some resource. 

The way that `influxdb` is written, it only listens on one port and that port will speak only one of HTTP or HTTPS at any given time and so it is not possible to downgrade from HTTPS to HTTP or to serve content over HTTP when the server is serving HTTPS.

Thinking about this wrt HSTS, when a TLS certificate IS NOT configured via https://docs.influxdata.com/influxdb/v2.7/security/enable-tls/, then `influxdb` will listen for HTTP traffic, however in this case HSTS cannot help because the `influxdb` server will not speak HTTPS even if the browser tried to upgrade to it (but also, `influxdb` listens on port `8086` by default, so the browser would instead go to port 443 if it did try to upgrade, which `influxdb` is not listening on. The `Strict-Transport-Security` header has no facility for saying what port to use for HTTPS). Conversely, when a TLS certificate IS configured, then `influxdb` will only listen for HTTPS traffic and attempting to speak HTTP with `influxdb` will result in a 400 error (bad request). As such, the portion of a browser's implementation of HSTS that converts HTTP requests to HTTPS is not particularly helpful.

The portion of a browser's implementation of HSTS that causes the browser to refuse to connect via HTTP could arguably be viewed as a hardening measure for installations who want to prevent contacting the `influxdb` server in the event that the administrator has disabled TLS. But thinking through this, let's assume for a moment that the administrator used `--http-bind-address` to set the port to 443 to workaround the `Strict-Transport-Security` port limitation; sending the `Strict-Transport-Security` unconditionally would be problematic for sites that need to temporarily disable TLS for some reason as an unexpired `max-age` could cause users to be unable to connect via HTTP in this temporary situation (this could easily happen during initial deployments that are troubleshooting during bootstrapping). A way around this would be when disabling TLS to also specify a different port or before disabling TLS to set `Strict-Transport-Security` with a `max-age` of `0`, have the users connect via HTTPS to reset the counter, and only then disable TLS. Both are cumbersome and therefore would have to be opt-in via configuration (eg, via `hardening-enabled` or some other option). BUT having it be opt-in doesn't help the problem of protecting against a changing configuration as the configuration could just as easily change the HSTS configuration as change the TLS configuration.

**Comment by jdstrand** [comment]
> * Threat 2 - server directs client to HTTP MiTM
>     * Question: how does the influx team ensure that the client doesn't act on HTTP redirects?

The server cannot inadvertently serve content over HTTP when configured with a TLS cert. InfluxDB is written in go and different listeners are created depending on if TLS is configured or not and it is not possible to have two listeners.

> * Threat 3 - MiTM serves up invalid cert
>     * Question: how does the influx team ensure that the user doesn't act on an invalid certificate from the MiTM?

It's true we rely on the browser to do the right thing here. Assuming the administrator of `influxdb` setup a valid certificate, then the browser should refuse to connect to the site in the event of a MITM with an invalid certificate and most browsers have made this sufficiently difficult to override. As you pointed out, HSTS makes it impossible for the user to override which is an added benefit over merely protecting against a changing configuration. We could consider adding an `--tls-hsts-max-age` option that would add this additional hardening. It would also be possible to include a default value with `hardening-enabled`. I'll let the team decide if they want to perform this feature enhancement and at what priority.

Fyi, I've also corrected myself in my previous comment after reading https://datatracker.ietf.org/doc/html/rfc6797#section-8.3: "_if the URI contains an explicit port component that is not equal to "80", the port component value MUST be preserved_"; `influxdb` listening on port 8086 should not be a problem.

Lastly, the tone of the conversation feels somewhat antagonistic, but perhaps that is just me. I can say that IMHO, adding an optional HSTS flag is possible and would achieve a marginal security hardening benefit, but a lack of this functionality does not "violate the claim that the instance is secure". When configured with TLS, the server will not serve HTTP content and the browser will alert the user to certificate errors.

**Comment by 3BK** [comment]
Hi Jamie.  Thanks for investigating. I'm sure there are a number of corner-cases which we haven't considered. ;)

**Comment by hartmafk** [comment]
@jdstrand 

I'm running into this issue now at work as well.

I agree that the real vulnerability is most likely pretty slim, however, HSTS has made its way into default profiles for vul scanners used to police big networks.

Having the option to enable HSTS would dramatically increase the user experience in this case, because a whole lot of arguing or excemtion processes would no longer be necessary.

**Comment by 3BK** [comment]
Agreed.  The HSTS issue undermines ease of use and may cause others to think twice.  Especially now, given that there are more viable low-cost competitors in the time series data historian niche.

________________________________

**Comment by 3BK** [comment]
In the past influx's biggest competitor was RRDtool.

Now, it's competitors include but aren't limited to VictoriaMetrics and Volt Data.  (Assuming we ignore sleepycat, HDF5, mongodb and the like.)
________________________________

**Comment by 3BK** [comment]
To tie a bow on this, one can wrap influxdb in a Web Application Firewall like NGINX; however, one might question - why all the friction?
________________________________

**Comment by devanbenz** [comment]
We've added a fix for this issue: https://github.com/influxdata/influxdb/pull/27489. Closing.
